From 1453f1ff9e5dcedb3c319c77656453a345aa2617 Mon Sep 17 00:00:00 2001
From: Julian Foad <julianfoad@apache.org>
Date: Mon, 10 Jun 2019 16:02:23 +0000
Subject: [PATCH] In svnserve, consistently handle errors in opening a
 repository.

These errors are still logged and reported to the client, as they were
before, but now it is done in the same way as everywhere else. (The error
logging now happens higher up the call stack.)

* subversion/svnserve/serve.c
  After reporting an error to the client, don't log it explicity here and
  then clear it; instead return it so the caller can do so.

git-svn-id: https://svn.apache.org/repos/asf/subversion/trunk@1860958 13f79535-47bb-0310-9956-ffa450edef68

Upstream status: backport(https://github.com/apache/subversion/commit/671a38245b05f6200c15fd24caf4fa1e9ee10cc5)
CVE: CVE-2019-0203

Signed-off-by: Anand Je Sypureddy <anandje@mvista.com>
---
 subversion/svnserve/serve.c | 13 +++++++------
 1 file changed, 7 insertions(+), 6 deletions(-)

diff --git a/subversion/svnserve/serve.c b/subversion/svnserve/serve.c
index 5192e7c..6159e22 100644
--- a/subversion/svnserve/serve.c
+++ b/subversion/svnserve/serve.c
@@ -4101,7 +4101,7 @@ construct_server_baton(server_baton_t **baton,
                        serve_params_t *params,
                        apr_pool_t *scratch_pool)
 {
-  svn_error_t *err, *io_err;
+  svn_error_t *err;
   apr_uint64_t ver;
   const char *client_url, *ra_client_string, *client_string;
   svn_ra_svn__list_t *caplist;
@@ -4239,11 +4239,12 @@ construct_server_baton(server_baton_t **baton,
     }
   if (err)
     {
-      log_error(err, b);
-      io_err = svn_ra_svn__write_cmd_failure(conn, scratch_pool, err);
-      svn_error_clear(err);
-      SVN_ERR(io_err);
-      return svn_ra_svn__flush(conn, scratch_pool);
+      /* Report these errors to the client before closing the connection. */
+      err = svn_error_compose_create(err,
+              svn_ra_svn__write_cmd_failure(conn, scratch_pool, err));
+      err = svn_error_compose_create(err,
+              svn_ra_svn__flush(conn, scratch_pool));
+      return err;
     }
 
   SVN_ERR(svn_fs_get_uuid(b->repository->fs, &b->repository->uuid,
-- 
2.17.1

